File: lib/services/frequencia_service.dart
import 'package:fluire/modelos/frequencia.dart';
import 'package:fluire/services/api_client.dart';

class FrequenciaService {
  final ApiClient _api = ApiClient.instance;

  Future<List<Frequencia>> listar() async {
    final response = await _api.get('/frequencias');

    if (response.statusCode != 200) {
      throw Exception('Erro ao carregar frequências: ${response.statusCode}');
    }

    final jsonData = ApiClient.decodificarCorpo(response);
    if (jsonData is! List) return [];

    return jsonData
        .whereType<Map<String, dynamic>>()
        .map((json) => Frequencia.fromJson(json))
        .toList();
  }

  Future<List<Frequencia>> listarPorAula(int aulaId) async {
    final response = await _api.get('/frequencias/aulas/$aulaId');

    if (response.statusCode != 200) {
      throw Exception('Erro ao carregar frequências da aula: ${response.statusCode}');
    }

    final jsonData = ApiClient.decodificarCorpo(response);
    if (jsonData is! List) return [];

    return jsonData
        .whereType<Map<String, dynamic>>()
        .map((json) => Frequencia.fromJson(json))
        .toList();
  }

  Future<Frequencia> registrar({
    required int aulaId,
    required int alunoId,
    required int presente,
    required String dataPresenca,
  }) async {
    final response = await _api.post('/frequencias', body: {
      'aula_id': aulaId,
      'aluno_id': alunoId,
      'presente': presente,
      'data_presenca': dataPresenca,
    });

    if (response.statusCode != 201) {
      final jsonData = ApiClient.decodificarCorpo(response);
      throw Exception(ApiClient.extrairErro(jsonData, fallback: 'Erro ao registrar frequência'));
    }

    final jsonData = ApiClient.decodificarCorpo(response);
    if (jsonData is Map<String, dynamic> && (jsonData.containsKey('aula_id') || jsonData.containsKey('id'))) {
      return Frequencia.fromJson(jsonData);
    }
    return Frequencia(
      aulaId: aulaId,
      alunoId: alunoId,
      presente: presente,
      dataPresenca: dataPresenca,
    );
  }
}

File: src/services/frequenciaService.js
import api, { extrairErro } from './api';

export async function listar() {
  const response = await api.get('/frequencias');
  const data = response.data;
  if (!Array.isArray(data)) return [];
  return data.filter((item) => item && typeof item === 'object').map(parseFrequencia);
}

export async function listarPorAula(aulaId) {
  const response = await api.get(`/frequencias/aulas/${aulaId}`);
  const data = response.data;
  if (!Array.isArray(data)) return [];
  return data.filter((item) => item && typeof item === 'object').map(parseFrequencia);
}

export async function registrar({ aulaId, alunoId, presente, dataPresenca }) {
  try {
    const response = await api.post('/frequencias', {
      aula_id: aulaId,
      aluno_id: alunoId,
      presente,
      data_presenca: dataPresenca,
    });
    const data = response.data;
    if (data && (data.aula_id || data.id)) {
      return parseFrequencia(data);
    }
    return { aulaId, alunoId, presente, dataPresenca };
  } catch (error) {
    throw new Error(extrairErro(error, 'Erro ao registrar frequência'));
  }
}

function parseFrequencia(json) {
  // presente vem como boolean do PostgreSQL (true/false), converte para 1/0
  const presente = json.presente === true || json.presente === 1 || json.presente === '1' ? 1 : 0;
  return {
    id: json.id?.toString() || null,
    aulaId: parseInt(json.aula_id, 10) || 0,
    alunoId: parseInt(json.aluno_id, 10) || 0,
    presente,
    dataPresenca: (json.data_presenca || json.dataPresenca || '').toString(),
  };
}
